File: DualKeyEncryption.cpp
#include "stdafx.h"
#include "DualKeyEncryption.h"


DualKeyEncryption::DualKeyEncryption(BigInteger semi_p)
{
	semi_prime = semi_p;
}

/*
* STEP VI:
*/
vector<string> DualKeyEncryption::encodeWord(vector<char> msg, BigInteger key) {
	vector<string> encryption = {};
	int x;
	for (int i = 0; i < msg.size(); i++) {
		// get unicode of this character as x
		x = (int)msg[i];
		// use RSA to encrypt & save in base 16
		encryption.push_back(BigInteger::valueOf(x).pow(key.intValue()).mod(semi_prime).toString(16));
	}

	return encryption;
}

/*
* STEP VII:
*/
string DualKeyEncryption::decodeWord(vector<string> code, BigInteger key) {
	string decryption = "";
	string char_string;
	int c;
	for (int i = 0; i < code.size(); i++) {
		char_string = "";
		// use RSA to decrypt
		c = BigInteger(code[i], 16).pow(key.intValue()).mod(semi_prime).intValue();
		char_string += (char)c;
		decryption += char_string;
	}

	return decryption;
}


DualKeyEncryption::~DualKeyEncryption()
{
}

File: DualKeyEncryption.java
package miscellaneous;

import java.math.BigInteger;

public class DualKeyEncryption {

    BigInteger semi_prime;

    public DualKeyEncryption(BigInteger semi_prime) {
        this.semi_prime = semi_prime;
    }

    /*
     * STEP VI:
     */
    public String[] encodeWord(char[] msg, BigInteger key) {
        String[] encryption = new String[msg.length];
        int x;
        for (int i = 0; i < msg.length; i++) {
            // get unicode of this character as x
            x = (int) msg[i];
            // use RSA to encrypt & save in base 16
            encryption[i] = BigInteger.valueOf(x).modPow(key, semi_prime).toString(16);
        }

        return encryption;
    }

    /*
     * STEP VII:
     */
    public String decodeWord(String[] code, BigInteger key) {
        String decryption = "";
        int c;
        for (int i = 0; i < code.length; i++) {
            // use RSA to decrypt
            c = (new BigInteger(code[i], 16)).modPow(key, semi_prime).intValue();
            decryption += Character.toString((char) c);
        }

        return decryption;
    }
}
